nca/Parsers/CalicoPolicyYamlParser.py

```diff
@@ -364,32 +392,61 @@ def _parse_icmp(self, icmp_data, not_icmp_data):
         allowed_keys = {'type': 0, 'code': 0}
         if icmp_data is not None:
             self.check_fields_validity(icmp_data, 'ICMP', allowed_keys)
-            err = ICMPDataSet.check_code_type_validity(icmp_type, icmp_code)
+            err = self.check_icmp_code_type_validity(icmp_type, icmp_code)
             if err:
                 self.syntax_error(err, icmp_data)
         if not_icmp_data is not None:
             self.check_fields_validity(not_icmp_data, 'notICMP', allowed_keys)
-            err = ICMPDataSet.check_code_type_validity(not_icmp_type, not_icmp_code)
+            err = self.check_icmp_code_type_validity(not_icmp_type, not_icmp_code)
             if err:
                 self.syntax_error(err, not_icmp_data)
 
-        res = ICMPDataSet(icmp_data is None and not_icmp_data is None)
+        protocols = ProtocolSet.get_protocol_set_with_single_protocol(protocol)
+        base_peer_set = self.peer_container.get_all_peers_group()
+        conn_cube = ConnectivityCube(base_peer_set)
+        if icmp_type:
+            conn_cube.set_dim("icmp_type", icmp_type)
+            if icmp_code:
+                conn_cube.set_dim("icmp_code", icmp_code)
+        not_conn_cube = ConnectivityCube(base_peer_set)
+        if not_icmp_type:
+            not_conn_cube.set_dim("icmp_type", not_icmp_type)
+            if not_icmp_code:
+                not_conn_cube.set_dim("icmp_code", not_icmp_code)
+        opt_conn_cube = conn_cube.copy()
+        opt_not_conn_cube = not_conn_cube.copy()
+        if self.optimized_run != 'false':
+            opt_conn_cube.set_dims({"src_peers": src_pods, "dst_peers": dst_pods, "protocols": protocols})
+            opt_not_conn_cube.set_dims({"src_peers": src_pods, "dst_peers": dst_pods, "protocols": protocols})
+
+        opt_props = ConnectivityProperties.make_empty_props()
         if icmp_data is not None:
-            res.add_to_set(icmp_type, icmp_code)
+            res = ConnectivityProperties.make_conn_props(conn_cube)
+            if self.optimized_run != 'false':
+                opt_props = ConnectivityProperties.make_conn_props(opt_conn_cube)
             if not_icmp_data is not None:
                 if icmp_type == not_icmp_type and icmp_code == not_icmp_code:
-                    res = ICMPDataSet()
+                    res = ConnectivityProperties.make_empty_props()
```

Comment:
should also set `opt_props` as empty.

Reply:
Reorganized this code, now `opt_props `is initialized as empty.